### ISSUE TITLE
Seeing hang when running Falcon40B decode demo on 8 chips

### ISSUE THREAD

**Issue by TT-BrianLiu** [tenstorrent/tt-metal#6472]
Able to run causalLM test with 60 layers for 8 chips with program cache enabled. Now, I am seeing a hang when running 8 chip demo for decode. Doesn't seem to be specific to program cache.

Branch: `bliu/issue-6472`
Branch: `main`
Test: ```pytest --disable-warnings -q -s --input-method=cli --cli-input="How do I get better at piano"  models/demos/falcon40b/demo/demo.py```

**Comment by jliangTT** [comment]
how reproducible is this? if it is a demo related issue, who should investigate it?

**Comment by jliangTT** [comment]
is this considered a blocker? c.c. @uaydonat

**Comment by TT-BrianLiu** [comment]
Very reproducible. Sometimes it can generate one token before hanging, sometimes it hangs on first token.

**Comment by jliangTT** [comment]
Ok - this sounds bad. i will hand it off to @eyonland to help triage.

**Comment by eyonland** [comment]
See comments on #6388 as it might be related.   @TT-BrianLiu , have tried collecting data on these issues with `valgrind --tool=helgrind` or `valgrind --tool=memcheck --leak-check=yes`?

**Comment by TT-BrianLiu** [comment]
I don't think these will help. If I enable more debug prints (ie. slowing things down), I can see it pass sometimes. It's for sure a race somewhere.

**Comment by aliuTT** [comment]
Pushed some fixes to `bliu/issue-6472`, decode demo on 8 chips passing for me on t3005. Please try. @TT-BrianLiu

**Comment by TT-BrianLiu** [comment]
I am running and confirming

**Comment by TT-BrianLiu** [comment]
It hung for me at 91st token for 8 chips 60 layers in demo

**Comment by TT-BrianLiu** [comment]
Allan and I both see demo working on: `bliu/issue-6472`
- 1024 `max_seq_len` gives gibberish
- 128 `max_seq_len` is good

So something on `main` is causing demo to hang + there is weirdness with different `max_seq_len`.

**Comment by TT-BrianLiu** [comment]
Unassigning @aliuTT , I can see which commit on `main` is breaking.

**Comment by TT-BrianLiu** [comment]
@jliangTT We can de-escalate this issue by the way. We have branches where it works and llama sees to work on`main` as well. I am trying to figure out why on `main`, Falcon40B still hangs randomly, but it's not a P0 anymore.

**Comment by davorchap** [comment]
@SeanNijjar is this still a an issue?

**Comment by SeanNijjar** [comment]
@davorchap - I'm aware of this hang: https://github.com/tenstorrent-metal/tt-metal/issues/6388, which is still reproducible in main. 

I'm not in the loop on this issue.

**Comment by TT-BrianLiu** [comment]
I am looking into it. We have one non-working branch (`main`) and one side branch working. Trying to figure out what the delta is...

**Comment by TT-BrianLiu** [comment]
I picked up as many commits as I can until demo is hanging. The commit that will cause it to hang is: bb9f6fc08020a33808c43bbe59f6a15354e0813e. 

Likely related to the hang seen in: #6388

**Comment by jliangTT** [comment]
@SeanNijjar, can you comment?

**Comment by SeanNijjar** [comment]
Hey @jliangTT what info are you looking for? It seems highly likely it's a duplicate of #6388. Although my commit is the first one to start causing a hang, I suspect a dispatch/tunnelling issue, given the signatures + behaviour in #6388

**Comment by jliangTT** [comment]
spun up a thread and the next step is to continue investigation in #6388. No more action pending on this one. Will keep it in P1 for now.

**Comment by SeanNijjar** [comment]
FYI I disabled bidirectional all-gather in main which instantiates fewer worker cores (and seemingly by extension), reduces the likelihood of the hang in #6388, so we should try this case again on main.

**Comment by jliangTT** [comment]
Do we know if this still hangs in main? @uaydonat

**Comment by uaydonat** [comment]
I tried it yesterday on sjc-nmrk-t3001 off main commit: 0df67e8af8 (Apr 12) and it PASSED with default arguments.

`pytest --disable-warnings -q -s --input-method=cli --cli-input="How do I get better at piano"  models/demos/falcon40b/demo/demo.py`

We should investigate more but this is a good sign.

**Comment by uaydonat** [comment]
max_seq_length=128 passed 3 times in a row, however, still seeing a hang on max_seq_length=1024

max_seq_length=1024 passed 9/10 times. In one run, it hung on token 529

**Comment by skhorasganiTT** [comment]
I am not seeing hangs on latest main (02550018ec4a27a33fcaf6630968b345e511a3f3) after upgrading to the new firmware running at 1GHz (made available in 13193c1) after running 100 times on sjc-snva-t3018 with max_seq_len=128.

**Comment by jliangTT** [comment]
how about max_seq_length=1024?

**Comment by skhorasganiTT** [comment]
max_seq_len=128 was also hanging (see #7836). I just ran the same prompt with max_seq_len=1024 10 times in a row, output looked okay and did not hang.

**Comment by uaydonat** [comment]
this is with prefill on device and with latest prefill optimizations? so a replicate of https://github.com/tenstorrent/tt-metal/issues/7836 ?

**Comment by skhorasganiTT** [comment]
Yes, we should probably close this and track there instead.

**Comment by uaydonat** [comment]
Duplicate of #7836
